### PR TITLE
Make an experimental single file build for the CLI

### LINKED ISSUE #221: Evaluate compiling/bundling for creating binary

### PR DESCRIPTION
**Pull request by zaiste**
Fixes #221

### CODE REVIEW

**Review by 2can — COMMENTED** [review]
@zaiste the only error I've found is for `info` command

`ERROR ENOENT: no such file or directory, open '/Users/tukan/Workspace/saleor/saleor-cli/fonts/Standard.flf'`

**Comment by lkostrowski** [comment]
Please wait with this, I don't want to introduce any instability when we don't have tests.

**Comment by zaiste** [comment]
@lkostrowski we can try to release it **only** as RC, and then revert if needed WDYT?

File: package.json [hunk 1]
```diff
@@ -7,33 +7,25 @@
...
   "bin": {
-    "saleor": "./build/cli.js"
+    "saleor": "./dist/saleor.js"
   },
   "scripts": {
-    "prepublishOnly": "tsc",
+    "prepublishOnly": "rm -rf ./package && pnpm build && clean-publish --fields dependencies",
+    "postpublish": "",
     "compile": "tsc",
+    "bundle": "esbuild src/cli.ts --bundle --minify --outfile=dist/saleor.js --platform=node --format=esm --target=node18 --banner:js=\"import { createRequire } from 'module';const require = createRequire(import.meta.url);import { dirname } from 'path'; import { fileURLToPath } from 'url'; const __dirname = dirname(fileURLToPath(import.meta.url));\" --out-extension:.js=.js",
```
Comment (lkostrowski) [inline package.json:16]: `--target=node18` I'd like to support at least node 16, but let's try with node 14 too. AFAIK we only need to polyfill or drop replaceAll api

File: src/cli.ts [hunk 2]
```diff
@@ -39,16 +39,14 @@ import {
...
-  pkg,
-  updateCheckInterval: 1000 * 60 * 15, // 15 minutes
-});
-notifier.notify({ isGlobal: true });
+// TODO temp disabled for the build
+// const notifier = updateNotifier({
+//   pkg: packageJson,
+//   updateCheckInterval: 1000 * 60 * 15, // 15 minutes
+// });
+// notifier.notify({ isGlobal: true });
```
Comment (lkostrowski) [inline src/cli.ts:49]: Why this is left like this?

**Review by lkostrowski — CHANGES_REQUESTED** [review]
I checked locally and it's great that CLI is <10mb. It seems that we don't need a binary and proper compilation is enough

However:
1. Let's support at least Node 16 (like before), but I hope we can include 14x too
2. Ensure no *behavior* change is introduced. This is a compilation so we can't merge any code that was commented out here.
3. We need to run QA test and ensure everything works. This change is simple, but swapping compiler can alter things we don't see. I'd like @michalina-graczyk to test it on branch and confirm there are no regressions

@michalina-graczyk to build it:
1. Clone repo and move to `make-esbuild` branch
2. Run `pnpm build` 
3. Run CLI as `node ./dist/saleor.js CMD`

**Comment by michalina-graczyk** [comment]
I have checked that. This is a list of my results:
- [x]  login
- [ ]  info

```jsx
saleor-cli git:(make-esbuild) ✗ node ./dist/saleor.js info

                                               ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
                                            ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
                ▄█████████████████████████████████████████▀░░░░░░░░░░░░░░░░
             ▄███▀                 ░░░░░░░░░░░░░░░ ▄███▀░░░░░░░░░░░░░░░
          ▄███▀                ░░░░░░░░░░░░░░░░░▄███▀░░░░░░░░░░░░░░
       ▄███▀               ░░░░░░░░░░░░░░░░░░▄███▀░░░░░░░░░░░░
    ▄███▀                                 ▄███▀
 ▄█████████████████████████████████████████▀         Saleor Commerce CLI v1.12.0

 ERROR ENOENT: no such file or directory, open '/Users/michalinqa/Desktop/projekty/CLI-253/saleor-cli/fonts/Standard.flf'
```

- [x]  node ./dist/saleor.js env create
- [x]  check if env is present on dasboard (check in graphical UI)
- [ ]  Deploy our react-storefront starter pack to Vercel

Untitled

- [ ]  create store front (saleor storefront create)

Untitled

- [ ]  check if dir with storefront project was created under selected name (or defalut if no name was picked "saleor-demo")
- [x]  check data in configuration file (.env in storefront folder, looking for NEXT_PUBLIC_API_URI) is connected to selected environment
- [x]  check if storefront is launched on listed port
- [ ]  execute "saleor storefront create --demo”

same result as stf create

- [ ]  check if demo project was created
- [ ]  check if demo is launched on listed port
- [x]  remove env (saleor env remove [key to env])
- [x]  logout
- [ ]  app create

Untitled

- [ ]  vercel login

Untitled

- [ ]  saleor app deploy on vercel

Untitled

File: package.json (hunk 1 above)
Comment (zaiste) [reply package.json]: Done

File: src/cli.ts (hunk 2 above)
Comment (zaiste) [reply src/cli.ts]: Fixed

**Comment by zaiste** [comment]
@michalina-graczyk @2can could you re-do your tests? it should work now

**Review by lkostrowski — APPROVED** [review]
Lets merge and branch v1.15 already

@michalina-graczyk please test on v1.15rc1 (diversely)

File: package.json [hunk 3]
```diff
@@ -22,19 +24,9 @@
     "release": "release-it --dry-run"
   },
   "files": [
-    "build",
```
Comment (mmiszy) [inline package.json:25]: Shouldn't there be `dist` instead?

File: src/cli/info.ts [hunk 4]
```diff
@@ -4,10 +4,10 @@ import Debug from 'debug';
 import { createRequire } from 'module';
 import type { CommandBuilder } from 'yargs';
 
+import packageJson from '../../package.json';
```
Comment (mmiszy) [inline src/cli/info.ts]: I haven't tested it, but does it work right now? Shouldn't it be:
```suggestion
import packageJson from '../../package.json' assert { type: 'json' };
```

Comment (zaiste) [reply src/cli/info.ts]: @mmiszy you're too quick! :) before you commented there's a new commit to disable it

File: package.json (hunk 3 above)
Comment (zaiste) [reply package.json:25]: @mmiszy yes, good catch

**Pull request merged**